src/app/component/Field/Field.component.js

```diff
@@ -146,6 +149,60 @@ class Field extends Component {
         }
     }
 
+    handleSelectListKeyPress(event) {
```

Comment:
I would conciser such implementation: 
```js
_getSelectedValueIndex(keyCode) {
        const { selectOptions } = this.props;
        const {
            searchString: prevSearchString,
            valueIndex: previousValueIndex
        } = this.state;

        const pressedKeyValue = String.fromCharCode(keyCode).toLowerCase();

        const searchString = (prevSearchString[prevSearchString.length - 1].toLowerCase() !== pressedKeyValue)
            ? `${prevSearchString}${pressedKeyValue}`
            : pressedKeyValue.toUpperCase();

        const nextValueIndex = selectOptions.findIndex(({ label }, i) => (
            label && label.startsWith(searchString) && (
                i > previousValueIndex || prevSearchString !== searchString
            )
        ));

        if (nextValueIndex !== -1) {
            return { searchString, valueIndex: nextValueIndex };
        }

        // if no items were found, take only the latest letter of the search string
        const newItemIndex = selectOptions.findIndex(({ label }) => (
            label && label.startsWith(
                searchString[searchString.length - 1].toUpperCase()
            )
        ));

        if (newItemIndex !== -1) {
            return { searchString, valueIndex: newItemIndex };
        }

        // if there are no items starting with this letter
        return {};
   }

   handleSelectListKeyPress(event) {
        const { selectOptions, onChange, id: selectId } = this.props;
        const { isSelectExpanded } = this.state;
        const keyCode = event.which || event.keycode;

        // on Enter pressed
        if (keyCode === 13) {
            this.handleSelectExpand();
            return;
        }

        if (!isSelectExpanded
            || !keyCode
            || keyCode < 65
            || keyCode > 122
            || (keyCode > 90 && keyCode < 97)
        ) return;

        const { searchString, valueIndex } = this._getSelectedValueIndex(keyCode);

        if (!searchString || !valueIndex) return;

        this.setState({ searchString, valueIndex }, () => {
            const { id, value } = selectOptions[valueIndex];
            // converting to string for avoiding the error with the first select option
            onChange(value.toString(10));
            const selectedElement = document.querySelector(`#${selectId} + ul #o${id}[role="menuitem"]`);
            selectedElement.focus();
        });
    }
```

Check if it is more readable to you.